### PR TITLE
find: Add exclude by patterns param.

### PR DESCRIPTION
**Pull request by k0ste**
Add ability to exclude from results some patterns. Regex supported.
It's useful when:
- need find all dirs in /home (user dirs), but need exclude some "service" dirs;
- find by pattern files, but exclude some unnecessary files (also may be useful with compare param), example:

``` yaml
---
- name: test
  become: false
  gather_facts: fasle
  remote_user: ansible
  no_log: false
  strategy: free
  hosts: localhost
  tasks:
  - name: "Step 1"
    find:
      paths: "/var/log"
      recurse: "no"
      file_type: "file"
      patterns: '^(.*)\.log$'
      excludes: '^(pacman|ntp)\.log$'
      use_regex: True
    register: "find_result"
  - name: "Step 2"
    debug:
      var: vars['find_result']['files']
```

### CODE REVIEW

**Comment by robynbergeron** [comment]
Thanks @k0ste for this PR. This module is maintained by the Ansible core team, so it can take a while for patches to be reviewed. Thanks for your patience.

File: files/find.py [hunk 1]
```diff
@@ -52,6 +51,14 @@
...
               least one of the patterns specified. Multiple patterns can be specified using a list.
         aliases: ['pattern']
+    excludes:
+        required: false
+        default: '*'
+        description:
+            - One or more (shell or regex) patterns, which type is controled by C(use_regex) option.
+            - Excludes is a patterns should not be returned in list. Multiple patterns can be specified
+              using a list.
+        aliases: ['exclude']
```
Comment (resmo) [inline files/find.py:63]: please add `version_added: "2.2"` here below aliases

**Comment by resmo** [comment]
Thanks @k0ste. Would you please rebase this PR rebase to devel.

**Comment by k0ste** [comment]
@resmo, done.

**Comment by resmo** [comment]
LGTM, the build fail is not related. @k0ste could you rebase again to get rid of the build failure, so this can be marked for inclusion?

**Comment by k0ste** [comment]
@resmo, do rebase. Or I need sync with devel? How I can do it "best way"?

**Comment by gregdek** [comment]
Thanks @k0ste for this PR. This PR requires revisions, either because it fails to build or by reviewer request. Please make the suggested revisions. When you are done, please comment with text 'ready_for_review' and we will put this PR back into review.

[This message brought to you by your friendly Ansibull-bot.]

**Comment by resmo** [comment]
Yes, you need to rebase on top of the current devel, e.g.:

```
git checkout devel
git pull git@github.com:ansible/ansible-modules-core.git devel
git checkout newfeature
git rebase devel
git push -f origin newfeature
```

File: files/find.py [hunk 2]
```diff
@@ -30,7 +30,7 @@
 ---
 module: find
 author: Brian Coca (based on Ruggero Marchei's Tidy)
-version_added: "2.0"
+version_added: "2.2"
```
Comment (resmo) [inline files/find.py]: do not change version_added here, but...

**Comment by k0ste** [comment]
ready_for_review

Big thanks @resmo

**Comment by resmo** [comment]
@k0ste one question, it looks like the way it is implemented, `pattern` and `exlclude` are mutually exclusive. IMHO it would be possible to use both right?  Not sure if this makes things more reliable though. Any thoughts, why you made them mutually exclusive?

If you want to keep them mutually exclusive, we should add `mutually_exclusive=[['pattern', 'exlclude']],` right before `supports_check_mode=True,`

**Comment by k0ste** [comment]
@resmo, `pattern` and `exclude` can work together.

See example on 1st topic.

**Comment by gregdek** [comment]
Thanks @k0ste for this PR. This module is maintained by the Ansible core team, so it can take a while for patches to be reviewed. Thanks for your patience.

Core team: please review according to guidelines (http://docs.ansible.com/ansible/developing_modules.html#module-checklist) and comment with 'needs_revision' or merge as appropriate.

[This message brought to you by your friendly Ansibull-bot.]

**Comment by resmo** [comment]
shipit (with 2.3)

**Comment by gregdek** [comment]
Thanks again to @k0ste for this PR. This PR was reviewed by an Ansible member. Marking for inclusion.

However, we would still appreciate if the maintainer(s) @ansible could look over it.

[This message brought to you by your friendly Ansibull-bot.]

**Comment by gregdek** [comment]
Thanks @k0ste for this PR. Unfortunately, it is not mergeable in its current state due to merge conflicts. Please rebase your PR. When you are done, please comment with text 'ready_for_review' and we will put this PR back into review.

For help on how to do this cleanly please see http://docs.ansible.com/ansible/community.html#contributing-code-features-or-bugfixes

[This message brought to you by your friendly Ansibull-bot.]

**Comment by k0ste** [comment]
ready_for_review

**Comment by gregdek** [comment]
Thanks @k0ste for this PR. This module is maintained by the Ansible core team, so it can take a while for patches to be reviewed. Thanks for your patience.

Core team: please review according to guidelines (http://docs.ansible.com/ansible/developing_modules.html#module-checklist) and comment with 'needs_revision' or merge as appropriate.

[This message brought to you by your friendly Ansibull-bot.]

**Comment by gregdek** [comment]
Thanks @k0ste for this PR. Unfortunately, it is not mergeable in its current state due to merge conflicts. Please rebase your PR. When you are done, please comment with text 'ready_for_review' and we will put this PR back into review.

For help on how to do this cleanly please see http://docs.ansible.com/ansible/community.html#contributing-code-features-or-bugfixes

[This message brought to you by your friendly Ansibull-bot.]

**Comment by k0ste** [comment]
ready_for_review

**Comment by gregdek** [comment]
Thanks @k0ste for this PR. This module is maintained by the Ansible core team, so it can take a while for patches to be reviewed. Thanks for your patience.

Core team: please review according to guidelines (http://docs.ansible.com/ansible/developing_modules.html#module-checklist) and comment with 'needs_revision' or merge as appropriate.

[This message brought to you by your friendly Ansibull-bot.]

**Comment by gregdek** [comment]
Thanks @k0ste for this PR. Unfortunately, it is not mergeable in its current state due to merge conflicts. Please rebase your PR. When you are done, please comment with text 'ready_for_review' and we will put this PR back into review.

For help on how to do this cleanly please see http://docs.ansible.com/ansible/community.html#contributing-code-features-or-bugfixes

[This message brought to you by your friendly Ansibull-bot.]

**Comment by k0ste** [comment]
ready_for_review

**Comment by gregdek** [comment]
Thanks @k0ste for this PR. This module is maintained by the Ansible core team, so it can take a while for patches to be reviewed. Thanks for your patience.

Core team: please review according to guidelines (http://docs.ansible.com/ansible/developing_modules.html#module-checklist) and comment with 'needs_revision' or merge as appropriate.

[This message brought to you by your friendly Ansibull-bot.]

**Pull request closed without merging**